### PR TITLE
feat: add phase0 searchbar

### PR DESCRIPTION
**Pull request by cbachmeier**
- adds the phase 0 searchbar
- spacing for mobile is off as not all components have been added

### CODE REVIEW

File: src/components/NavBar/SearchBar.tsx [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+    toggleOpen()
+  }
+
+  // Close the modal on escape
+  useEffect(() => {
+    const keyDownHandler = (event: KeyboardEvent) => {
+      if (event.key === 'Enter' && isHovered) {
+        event.preventDefault()
+        // TODO update with correct token explore URI
+        navigate(`tokens/${token.address}`)
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:145]: you'll want to do `/tokens/...` with react-router-v6

File: src/components/NavBar/SearchBar.tsx [hunk 2]
```diff
@@ -0,0 +1,556 @@
...
+        event.preventDefault()
+        navigate(`/nft/collection/${collection.address}`)
+        handleClick()
+      }
+    }
+    document.addEventListener('keydown', keyDownHandler)
+    return () => {
+      document.removeEventListener('keydown', keyDownHandler)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:67]: think it's an anti-pattern to avoid this. was there a bug to list them all?

File: src/components/NavBar/SearchBar.tsx [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+  }, [toggleOpen, history, isHovered, collection])
+
+  return (
+    <Row
+      as="a"
+      href={`/#/nft/collection/${collection.address}`}
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:73]: we'll want to use `react-router` link here

File: src/components/NavBar/SearchBar.tsx [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+
+interface SearchBarDropdownProps {
+  toggleOpen: () => void
+  tokens: FungibleToken[]
+  collections: GenieCollection[]
+  hasInput: boolean
+}
+
+export const SearchBarDropdown = ({ toggleOpen, tokens, collections, hasInput }: SearchBarDropdownProps) => {
+  const [hoveredIndex, setHoveredIndex] = useState<number | undefined>(undefined)
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:295]: why is hoveredIndex needed vs using :hover?

File: src/components/NavBar/SearchBar.tsx [hunk 5]
```diff
@@ -0,0 +1,556 @@
...
+
+  const totalSuggestions = hasInput
+    ? tokens.length + collections.length
+    : Math.min(searchHistory.length, 2) +
+      (isNFTPage || !isTokenPage ? trendingCollections?.length ?? 0 : 0) +
+      (isTokenPage || !isNFTPage ? trendingTokens?.length ?? 0 : 0)
+
+  // Close the modal on escape
+  useEffect(() => {
+    const keyDownHandler = (event: KeyboardEvent) => {
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:351]: would love to move the keyboard hotkey stuff out since you have a few in here. perhaps a `useHotKeyHandler` hook?

File: src/components/NavBar/SearchBar.tsx [hunk 6]
```diff
@@ -0,0 +1,556 @@
...
+    ['searchTokens', debouncedSearchValue],
+    () => fetchSearchTokens(debouncedSearchValue),
+    {
+      refetchOnWindowFocus: false,
+      refetchOnMount: false,
+      refetchOnReconnect: false,
+    }
+  )
+
+  const isNFTPage = location.hash.includes('/nft')
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:464]: use `useLocation` for this

File: src/components/NavBar/SearchBar.tsx [hunk 7]
```diff
@@ -0,0 +1,556 @@
...
+
+export const SearchBarDropdown = ({ toggleOpen, tokens, collections, hasInput }: SearchBarDropdownProps) => {
+  const [hoveredIndex, setHoveredIndex] = useState<number | undefined>(undefined)
+  const searchHistory = useSearchHistory(
+    (state: { history: (FungibleToken | GenieCollection)[] }) => state.history
+  ).slice(0, 2)
+  const isNFTPage = location.hash.includes('/nft')
+  const isTokenPage = location.hash.includes('/token')
+
+  const tokenSearchResults =
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:302]: is there a way to move the business logic into a utils file in this folder? hard to follow the UI logic with the algorithm mixed in

File: src/components/NavBar/SearchBar.tsx [hunk 8]
```diff
@@ -0,0 +1,556 @@
...
+    {
+      refetchOnWindowFocus: false,
+      refetchOnMount: false,
+      refetchOnReconnect: false,
+    }
+  )
+
+  const isNFTPage = location.hash.includes('/nft')
+
+  // If not an nft page show up to 5 tokens, else up to 3. Max total suggestions of 8
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:466]: same comment here about business logic

File: src/components/NavBar/SearchBar.tsx (hunk 4 above)
Comment (cbachmeier) [reply src/components/NavBar/SearchBar.tsx:290]: A couple of reasons:
1. This allows us to track the user across the various sections in the suggestions dropdown including past search history, trending tokens, trending collection, token search results, and collection search results
2. This handles for both mouse and keyboard event tracking

File: src/components/NavBar/SearchBar.tsx (hunk 5 above)
Comment (cbachmeier) [reply src/components/NavBar/SearchBar.tsx:347]: We need some the keyboard logic to be localized to the suggestion row components, especially for tracking hovered/focused state, which is why we cant abstract to a generalized hook. You can read more on our decision when this was developed in this pr: https://github.com/Uniswap/interface-nft/pull/26

File: src/components/NavBar/SearchBar.tsx (hunk 7 above)
Comment (cbachmeier) [reply src/components/NavBar/SearchBar.tsx:298]: This case is just a conditional checking if there are search results otherwise show the blank state, I dont think a util function is necessary for this case

File: src/components/NavBar/SearchBar.tsx (hunk 8 above)
Comment (cbachmeier) [reply src/components/NavBar/SearchBar.tsx:462]: I've moved this to a util function

File: src/components/NavBar/SearchBar.tsx [hunk 9]
```diff
@@ -0,0 +1,546 @@
...
+  const searchRef = useRef<HTMLDivElement>(null)
+  const location = useLocation()
+  const { width: windowWidth } = useWindowSize()
+
+  useOnClickOutside(searchRef, () => {
+    isOpen && toggleOpen()
+  })
+
+  // TODO NFT Search Results implmented here
+  // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:448]: rm?

Comment (cbachmeier) [reply src/components/NavBar/SearchBar.tsx:448]: This is needed for initializing a blank collection array. This won't be needed when we've added collection search results

File: src/components/NavBar/SearchBar.tsx [hunk 10]
```diff
@@ -0,0 +1,546 @@
...
+  hasInput: boolean
+}
+
+export const SearchBarDropdown = ({ toggleOpen, tokens, collections, hasInput }: SearchBarDropdownProps) => {
+  const [hoveredIndex, setHoveredIndex] = useState<number | undefined>(undefined)
+  const searchHistory = useSearchHistory(
+    (state: { history: (FungibleToken | GenieCollection)[] }) => state.history
+  ).slice(0, 2)
+  const location = useLocation()
+  const isNFTPage = location.hash.includes('/nfts')
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:296]: [`useLocation`](https://v5.reactrouter.com/web/api/Hooks/uselocation)

Comment (vm) [inline src/components/NavBar/SearchBar.tsx:296]: can't we use path?

Comment (cbachmeier) [reply src/components/NavBar/SearchBar.tsx:296]: whats the advantage of using path over location hash?

File: src/components/NavBar/SearchBar.tsx [hunk 11]
```diff
@@ -0,0 +1,546 @@
...
+        startingIndex={isNFTPage ? 0 : tokens.length}
+        setHoveredIndex={setHoveredIndex}
+        toggleOpen={toggleOpen}
+        suggestions={collections}
+        header={'NFT Collections'}
+      />
+    ) : null
+
+  // TODO Trending NFT Results implmented here
+  const trendingCollections = [] as TrendingCollection[]
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:326]: i suggest moving this to a util called `getTrendingCollections`

File: src/components/NavBar/SearchBar.tsx [hunk 12]
```diff
@@ -0,0 +1,546 @@
...
+        event.preventDefault()
+        navigate(`/tokens/${token.address}`)
+        handleClick()
+      }
+    }
+    document.addEventListener('keydown', keyDownHandler)
+    return () => {
+      document.removeEventListener('keydown', keyDownHandler)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:151]: why the exhaustive deps ignore here?

File: src/components/NavBar/SearchBar.tsx [hunk 13]
```diff
@@ -0,0 +1,546 @@
...
+
+interface TokenRowProps {
+  token: FungibleToken
+  isHovered: boolean
+  setHoveredIndex: (index: number | undefined) => void
+  toggleOpen: () => void
+  index: number
+}
+
+const TokenRow = ({ token, isHovered, setHoveredIndex, toggleOpen, index }: TokenRowProps) => {
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:125]: move TokenRow to it's own file

File: src/components/NavBar/SearchBar.tsx [hunk 14]
```diff
@@ -0,0 +1,546 @@
...
+          <Box className={styles.secondaryText} color={token.price24hChange >= 0 ? 'green400' : 'red400'}>
+            {token.price24hChange.toFixed(2)}%
+          </Box>
+        )}
+      </Column>
+    </Link>
+  )
+}
+
+const SkeletonRow = () => {
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:211]: skeleton row can go in the tokenrow file if you want, or a separate file as well

File: src/components/NavBar/SearchBar.tsx (hunk 11 above)
Comment (cbachmeier) [reply src/components/NavBar/SearchBar.tsx:326]: this is temporary before we add in the query for fetching collection search results

File: src/components/NavBar/SearchBar.tsx [hunk 15]
```diff
@@ -0,0 +1,546 @@
...
+  toggleOpen: () => void
+  suggestions: (GenieCollection | FungibleToken)[]
+  header: string
+  headerIcon?: JSX.Element
+  hoveredIndex: number | undefined
+  startingIndex: number
+  setHoveredIndex: (index: number | undefined) => void
+}
+
+export const SearchBarDropdownSection = ({
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:241]: own file for this also

File: src/components/NavBar/SearchBar.tsx (hunk 10 above)
Comment (vm) [reply src/components/NavBar/SearchBar.tsx:296]: we may not use hash if we switch away from hashrouter, so that's implementation detail

File: src/components/NavBar/SearchBar.tsx (hunk 11 above)
Comment (vm) [reply src/components/NavBar/SearchBar.tsx:326]: that's a good reason to separate it imo, bc we would likely move most of the logic into a hook that's separated

Comment (vm) [reply src/components/NavBar/SearchBar.tsx:325]: optional

File: src/components/NavBar/SearchBar.tsx (hunk 13 above)
Comment (vm) [reply src/components/NavBar/SearchBar.tsx:125]: optional

File: src/components/NavBar/SearchBar.tsx (hunk 15 above)
Comment (vm) [reply src/components/NavBar/SearchBar.tsx:240]: optional

File: src/components/NavBar/SearchBar.css.ts [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+  }),
+  {
+    lineHeight: '20px',
+  },
+])
+
+export const imageHolder = style([
+  suggestionImage,
+  {
+    background: 'radial-gradient(167.86% 167.86% at -21.43% -50%, #4C82FB 0%, #09265E 100%)',
```
Comment (vm) [inline src/components/NavBar/SearchBar.css.ts:125]: could you confirm with fred about the hardcoded background values here?

File: src/components/NavBar/SearchBar.tsx [hunk 17]
```diff
@@ -0,0 +1,545 @@
...
+            as="img"
+            src={collection.imageUrl}
+            alt={collection.name}
+            className={styles.suggestionImage}
+            onError={() => setBrokenImage(true)}
+            onLoad={() => setLoaded(true)}
+            style={{
+              background: loaded
+                ? 'none'
+                : 'radial-gradient(167.86% 167.86% at -21.43% -50%, #4C82FB 0%, #09265E 100%)',
```
Comment (vm) [inline src/components/NavBar/SearchBar.tsx:91]: this seems to be the same background as in the sprinkles? can we make a variable for it

File: src/components/NavBar/SearchBar.tsx (hunk 5 above)
Comment (vm) [reply src/components/NavBar/SearchBar.tsx:347]: optional

File: src/components/NavBar/SearchBar.css.ts (hunk 16 above)
Comment (cbachmeier) [reply src/components/NavBar/SearchBar.css.ts:125]: this was taken directly from the Figma

File: src/components/NavBar/SearchBar.tsx (hunk 17 above)
Comment (cbachmeier) [reply src/components/NavBar/SearchBar.tsx]: made the classname a conditional to condense useage

File: src/lib/utils/searchBar.ts [hunk 18]
```diff
@@ -0,0 +1,20 @@
+import { FungibleToken, GenieCollection } from 'nft/types'
```
Comment (vm) [inline src/lib/utils/searchBar.ts:1]: nit: would bet better if this could live within the NavBar component dir imo

**Review by vm — APPROVED** [review]

File: src/lib/utils/searchBar.ts (hunk 18 above)
Comment (cbachmeier) [reply src/lib/utils/searchBar.ts:1]: We like to have all datatypes in the `types` folder for organization. GenieCollection is used in several places outside of the NavBar

File: src/components/NavBar/SearchBar.css.ts (hunk 16 above)
Comment (cbachmeier) [reply src/components/NavBar/SearchBar.css.ts]: Talked with design and they actually wanted to update this to be the same color as the skeleton loading state
